diff --git a/app/app/src/main/java/com/focusup/focusupapp/MainActivity.kt b/app/app/src/main/java/com/focusup/focusupapp/MainActivity.kt
index 05368a6..7736e70 100644
--- a/app/app/src/main/java/com/focusup/focusupapp/MainActivity.kt
+++ b/app/app/src/main/java/com/focusup/focusupapp/MainActivity.kt
@@ -62,6 +62,8 @@ import com.focusup.focusupapp.ui.theme.SpecialGreen
 import com.focusup.focusupapp.ui.theme.SpecialGreen2
 import com.focusup.focusupapp.ui.theme.SpecialBlue
 import com.focusup.focusupapp.ui.theme.SpecialBlue2
+import com.focusup.focusupapp.ui.theme.SpecialGray
+import com.focusup.focusupapp.ui.theme.SpecialGray2
 
 import android.app.NotificationChannel
 import android.app.NotificationManager
@@ -110,6 +112,9 @@ import androidx.compose.ui.geometry.Offset
 import androidx.compose.ui.geometry.Size
 import androidx.compose.material3.OutlinedTextFieldDefaults
 import androidx.compose.foundation.lazy.grid.rememberLazyGridState
+import androidx.compose.material.icons.filled.Delete
+import androidx.compose.ui.text.style.TextDecoration
+import com.focusup.focusupapp.model.Step
 
 // Crear canal de notificacion
 @RequiresApi(Build.VERSION_CODES.O)
@@ -536,7 +541,7 @@ fun CalendarScreen(context: Context) {
                                         Surface(
                                             tonalElevation = 0.dp,
                                             shape = RoundedCornerShape(6.dp),
-                                            color = GetTaskColor(task.difficulty),
+                                            color = GetTaskColor(task.difficulty, task.isCompleted),
                                             modifier = Modifier
                                                 .fillMaxWidth()
                                                 .padding(vertical = 2.dp)
@@ -580,7 +585,7 @@ fun CalendarScreen(context: Context) {
                     onDismissRequest = { showTaskDialog = false },
                     title = {
                         Text(
-                            text = selectedTask!!.title,
+                            text = selectedTask!!.title + getStatus(selectedTask!!.isCompleted),
                             color = MaterialTheme.colorScheme.onSecondary
                         )
                     },
@@ -648,7 +653,7 @@ fun CalendarScreen(context: Context) {
                                             }
                                         )
             
-                                        // Texto (con tachado si está completado)
+                                        // Texto (con tachado si esta completado)
                                         Text(
                                             text = "- ${step.text}",
                                             style = MaterialTheme.typography.bodySmall.copy(
@@ -698,6 +703,33 @@ fun CalendarScreen(context: Context) {
                     confirmButton = {
                         Button(
                             onClick = {
+                                // Si la tarea ya estaba completada, descompletarla
+                                if (selectedTask!!.isCompleted) {
+                                    val result = AccountStorage.uncompleteTaskOfAccount(context, currentAccount!!.id, selectedTask!!.id)
+                                    if (!result) {
+                                        selectedTask = null
+                                        showTaskDialog = false
+                                        if (isPremium) {
+                                            showTaskNotCreatedDialogPremium = true
+                                        } else {
+                                            showTaskNotCreatedDialogNonPremium = true
+                                        }
+                                        return@Button
+                                    }
+                                    val updatedAccount = SessionStorage.getUpdatedSession(context)
+                                    if (updatedAccount != null) {
+                                        currentAccount = updatedAccount
+                                        SessionStorage.refreshSession(context, updatedAccount)
+                                    }
+                                    // Actualizar la lista de tareas en la UI
+                                    val updatedTasks = AccountStorage.getTasksForAccount(context, currentAccount!!.id)
+                                    tasksList.clear()
+                                    tasksList.addAll(updatedTasks)
+
+                                    selectedTask = null
+                                    showTaskDialog = false
+                                    return@Button
+                                }
                                 showTaskDialog = false
                                 lastPointsEarned = 0
             
@@ -716,52 +748,48 @@ fun CalendarScreen(context: Context) {
                                         SessionStorage.refreshSession(context, updatedAccount)
                                     }
                                 }
-            
-                              // Marcar tarea como completa en lugar de eliminarla
-                            selectedTask?.let { task ->
-                            task.isCompleted = true  // <<-- marca la tarea como completa
-
-                            // Mover al final de la lista
-                            // Marcar tarea como completa y moverla al final de la lista si hay otras incompletas
-
-                            tasksList.remove(task)
-                            tasksList.add(task)
-
-                                if (currentAccount != null) {
-                                AccountStorage.updateTaskOfAccount(context, currentAccount!!.id, task)
-
-                                val updatedAccount = SessionStorage.getUpdatedSession(context)
-                                if (updatedAccount != null) {
-                                currentAccount = updatedAccount
-                                SessionStorage.refreshSession(context, updatedAccount)
-        }
-    }
-}
-
-            
+                                // Marcar tarea como completa en lugar de eliminarla
+                                selectedTask?.let { task ->
+                                    task.isCompleted = true
+                                    // Mover al final de la lista
+                                    // Marcar tarea como completa y moverla al final de la lista si hay otras incompletas
+                                    if (currentAccount != null) {
+                                        AccountStorage.updateTaskOfAccount(context, currentAccount!!.id, task)
+
+                                        val updatedAccount = SessionStorage.getUpdatedSession(context)
+                                        if (updatedAccount != null) {
+                                            currentAccount = updatedAccount
+                                            SessionStorage.refreshSession(context, updatedAccount)
+                                        }
+                                    }
+                                    // Actualizar la lista de tareas en la UI
+                                    val updatedTasks = AccountStorage.getTasksForAccount(context, currentAccount!!.id)
+                                    tasksList.clear()
+                                    tasksList.addAll(updatedTasks)
+                                }
                                 selectedTask = null
                                 showTaskCompletedDialog = true
                             },
                             colors = ButtonDefaults.buttonColors(
-                                containerColor = GetTaskColorText(selectedTask!!.difficulty),
+                                containerColor = GetTaskColorText(selectedTask!!.difficulty, selectedTask!!.isCompleted),
                                 contentColor = MaterialTheme.colorScheme.onSecondary
                             )
                         ) {
-                            Text("Completar tarea")
+                            Text(getCompleteButtonText(selectedTask!!.isCompleted))
                         }
                     },
                     dismissButton = {
                         Button(
                             onClick = { showTaskDialog = false },
                             colors = ButtonDefaults.buttonColors(
-                                containerColor = GetTaskColorText(selectedTask!!.difficulty),
+                                containerColor = GetTaskColorText(selectedTask!!.difficulty, selectedTask!!.isCompleted),
                                 contentColor = MaterialTheme.colorScheme.onSecondary
                             )
                         ) {
                             Text("Cerrar")
                         }
                     },
-                    containerColor = GetTaskColor(selectedTask!!.difficulty)
+                    containerColor = GetTaskColor(selectedTask!!.difficulty, selectedTask!!.isCompleted)
                 )
             }
             
@@ -786,11 +814,6 @@ fun CalendarScreen(context: Context) {
             
                         if (result) {
                             showTaskCreatedDialog = true
-                        }
-                    }
-                )
-            }
-            
                             // Actualizar la sesión y las tareas
                             val updatedAccount = SessionStorage.getUpdatedSession(context)
                             if (updatedAccount != null) {
@@ -878,6 +901,44 @@ fun CalendarScreen(context: Context) {
                 SettingsDialog(onDismiss = { showSettingsDialog = false })
             }
 
+            if (showLoginRequiredForTask) {
+                AlertDialog(
+                    onDismissRequest = { showLoginRequiredForTask = false },
+                    title = { Text("Sesión requerida", color = MaterialTheme.colorScheme.onBackground) },
+                    text = { 
+                        Text(
+                            "Debes iniciar sesión para poder crear tareas.", 
+                            color = MaterialTheme.colorScheme.onBackground
+                        ) 
+                    },
+                    confirmButton = {
+                        Button(
+                            onClick = { 
+                                showLoginRequiredForTask = false 
+                                showLoginDialog = true
+                            },
+                            colors = ButtonDefaults.buttonColors(
+                                containerColor = MaterialTheme.colorScheme.primary,
+                                contentColor = MaterialTheme.colorScheme.onPrimary
+                            )
+                        ) {
+                            Text("Iniciar sesión")
+                        }
+                    },
+                    dismissButton = {
+                        Button(
+                            onClick = { showLoginRequiredForTask = false },
+                            colors = ButtonDefaults.buttonColors(
+                                containerColor = MaterialTheme.colorScheme.secondary,
+                                contentColor = MaterialTheme.colorScheme.onSecondary
+                            )
+                        ) {
+                            Text("Cancelar")
+                        }
+                    },
+                    containerColor = MaterialTheme.colorScheme.surface
+                )
+            }
         }
     }
     if (showRewards) {
@@ -1362,7 +1423,7 @@ fun AddTaskDialog(
                                 dueDate = taskDueDate,
                                 dueTime = taskDueTime.withSecond(0).withNano(0),
                                 difficulty = selectedDifficulty,
-                                steps = taskSteps,
+                                steps = convertStringsToSteps(taskSteps),
                                 id = nextId
                             )
                             onAddTask(newTask)
@@ -1741,7 +1802,7 @@ fun TaskNotCreatedDialogNonPremium(
         onDismissRequest = onDismiss,
         title = { Text("Tarea no creada", color = MaterialTheme.colorScheme.onBackground) },
         text = {
-            Text("No se pudo crear la tarea. Ha alcanzado el límite de tareas.\nConsidera completar tareas existentes o usar una cuenta Premium para tener más tareas activas.",
+            Text("No se pudo crear/descompletar la tarea. Ha alcanzado el límite de tareas.\nConsidera completar tareas existentes o usar una cuenta Premium para tener más tareas activas.",
             color = MaterialTheme.colorScheme.onBackground)
         },
         confirmButton = {
@@ -1768,7 +1829,7 @@ fun TaskNotCreatedDialogPremium(
         onDismissRequest = onDismiss,
         title = { Text("Tarea no creada", color = MaterialTheme.colorScheme.onBackground) },
         text = {
-            Text("No se pudo crear la tarea. Ha alcanzado el límite de tareas.\nConsidera completar tareas existentes.",
+            Text("No se pudo crear/descompletar la tarea. Ha alcanzado el límite de tareas.\nConsidera completar tareas existentes.",
             color = MaterialTheme.colorScheme.onBackground)
         },
         confirmButton = {
@@ -1816,16 +1877,6 @@ fun TaskCompletedDialog(
         containerColor = MaterialTheme.colorScheme.surface
     )
 }
-//Funcion para mostrar los pasos tachados
-Text(
-    text = step.text,
-    style = TextStyle(
-        textDecoration = if (step.isCompleted)
-            TextDecoration.LineThrough
-        else
-            TextDecoration.None
-    )
-)
 
 
 // Funcion para mostrar el manual de usuario o tutorial
@@ -2050,9 +2101,12 @@ fun CalendarPreview() {
     }
 }
 
-// Funcion para obtener un color segun la dificultad
+// Funcion para obtener un color segun la dificultad y si la tarea esta completada
 @Composable
-fun GetTaskColor(difficulty: Int): Color {
+fun GetTaskColor(difficulty: Int, isCompleted: Boolean): Color {
+    if (isCompleted) {
+        return SpecialGray
+    }
     return when (difficulty) {
         5 -> SpecialRed
         4 -> SpecialOrange
@@ -2063,9 +2117,12 @@ fun GetTaskColor(difficulty: Int): Color {
     }
 }
 
-// Funcion para obtener un color para texto segun la dificultad
+// Funcion para obtener un color para texto segun la dificultad y si la tarea esta completada
 @Composable
-fun GetTaskColorText(difficulty: Int): Color {
+fun GetTaskColorText(difficulty: Int, isCompleted: Boolean): Color {
+    if (isCompleted) {
+        return SpecialGray2
+    }
     return when (difficulty) {
         5 -> SpecialRed2
         4 -> SpecialOrange2
@@ -2102,6 +2159,21 @@ fun CalculatePointsForTask(task: Task): Int {
     return basePoints + (limitedDaysEarly * difficulty)
 }
 
+// Funcion para volver una lista de strings a una lista de steps
+fun convertStringsToSteps(stepStrings: List<String>): MutableList<Step> {
+    return stepStrings.map { Step(it, false) }.toMutableList()
+}
+
+// Funcion para obtener el estatus de la tarea
+fun getStatus(isCompleted: Boolean): String {
+    return if (isCompleted) " (Completada)" else ""
+}
+
+// Funcion para obtener el texto de "Completar tarea" o "Descompletar tarea"
+fun getCompleteButtonText(isCompleted: Boolean): String {
+    return if (isCompleted) "Descompletar tarea" else "Completar tarea"
+}
+
 /*
 
 package com.example.focusup
diff --git a/app/app/src/main/java/com/focusup/focusupapp/model/Task.kt b/app/app/src/main/java/com/focusup/focusupapp/model/Task.kt
index 95a2f85..5bf7ff0 100644
--- a/app/app/src/main/java/com/focusup/focusupapp/model/Task.kt
+++ b/app/app/src/main/java/com/focusup/focusupapp/model/Task.kt
@@ -19,6 +19,6 @@ data class Task(
     var steps: MutableList<Step>,
     // ID
     val id: Int,
+    // Indica si la tarea esta completada
     var isCompleted: Boolean = false
-
 )
\ No newline at end of file
diff --git a/app/app/src/main/java/com/focusup/focusupapp/storage/AccountStorage.kt b/app/app/src/main/java/com/focusup/focusupapp/storage/AccountStorage.kt
index 0af7c3d..dde63a1 100644
--- a/app/app/src/main/java/com/focusup/focusupapp/storage/AccountStorage.kt
+++ b/app/app/src/main/java/com/focusup/focusupapp/storage/AccountStorage.kt
@@ -11,6 +11,7 @@ import java.time.LocalDate
 import java.time.LocalTime
 
 object AccountStorage {
+    private const val MAX_TASKS_NON_PREMIUM = 10
     private const val ACCOUNTS_FILE = "accounts.json"
     private val gson: Gson = GsonBuilder()
         .registerTypeAdapter(LocalDate::class.java, LocalDateAdapter())
@@ -106,11 +107,13 @@ object AccountStorage {
         val account = accounts[accountIndex]
         
         // Verificar límite de tareas según si es premium o no
-        if (!isPremium && account.tasks.size >= 5) {
+        val uncompletedTasksCount = account.tasks.count { !it.isCompleted }
+        if (!isPremium && uncompletedTasksCount >= MAX_TASKS_NON_PREMIUM) {
             return false
         }
-        
         account.tasks.add(task)
+        // Mover las tareas completadas al final de la lista
+        account.tasks = account.tasks.sortedWith(compareBy<Task> { it.isCompleted }).toMutableList()
         saveAccounts(context, accounts)
         return true
     }
@@ -187,4 +190,62 @@ object AccountStorage {
         val accounts = loadAccounts(context)
         return accounts.find { it.id == accountId }
     }
+
+    /**
+     * Actualiza el estado de un paso en una tarea específica
+     */
+    fun updateStepInTask(context: Context, accountId: Int, taskId: Int, stepIndex: Int,isCompleted: Boolean) {
+        val accounts = loadAccounts(context).toMutableList()
+        val accountIndex = accounts.indexOfFirst { it.id == accountId }
+        if (accountIndex == -1) return
+        val task = accounts[accountIndex].tasks.find { it.id == taskId } ?: return
+        if (stepIndex in task.steps.indices) {
+            task.steps[stepIndex].isCompleted = isCompleted
+            saveAccounts(context, accounts)
+        }
+    }
+
+    /**
+     * Actualiza una tarea específica en una cuenta
+     */
+    fun updateTaskOfAccount(context: Context, accountId: Int, updatedTask: Task) {
+        val accounts = loadAccounts(context).toMutableList()
+        val accountIndex = accounts.indexOfFirst { it.id == accountId }
+        if (accountIndex == -1) return
+        val account = accounts[accountIndex]
+        val taskIndex = account.tasks.indexOfFirst { it.id == updatedTask.id }
+        if (taskIndex == -1) return
+        // Actualizar la tarea
+        account.tasks[taskIndex] = updatedTask
+        // Si la tarea esta completada, moverla al final de la lista
+        if (updatedTask.isCompleted) {
+            val completedTask = account.tasks.removeAt(taskIndex)
+            account.tasks.add(completedTask)
+        }
+        saveAccounts(context, accounts)
+    }
+
+    /**
+     * Descompleta una tarea específica en una cuenta
+     */
+    fun uncompleteTaskOfAccount(context: Context, accountId: Int, taskId: Int): Boolean {
+        val accounts = loadAccounts(context).toMutableList()
+        val accountIndex = accounts.indexOfFirst { it.id == accountId }
+        if (accountIndex == -1) return false
+        val account = accounts[accountIndex]
+        val uncompletedTasksCount = account.tasks.count { !it.isCompleted }
+        val isPremium = account.isPremium
+        if (!isPremium && uncompletedTasksCount >= MAX_TASKS_NON_PREMIUM) {
+            return false
+        }
+        val taskIndex = account.tasks.indexOfFirst { it.id == taskId }
+        if (taskIndex == -1) return false
+        val task = account.tasks[taskIndex]
+        task.isCompleted = false
+        // Mover la tarea al inicio de la lista
+        val uncompletedTask = account.tasks.removeAt(taskIndex)
+        account.tasks.add(0, uncompletedTask)
+        saveAccounts(context, accounts)
+        return true
+    }
 }
\ No newline at end of file
diff --git a/app/app/src/main/java/com/focusup/focusupapp/ui/theme/Color.kt b/app/app/src/main/java/com/focusup/focusupapp/ui/theme/Color.kt
index 8f430ff..31a10dc 100644
--- a/app/app/src/main/java/com/focusup/focusupapp/ui/theme/Color.kt
+++ b/app/app/src/main/java/com/focusup/focusupapp/ui/theme/Color.kt
@@ -125,3 +125,5 @@ val SpecialGreen = Color(0xFF9DF89D)
 val SpecialGreen2 = Color(0xFF5CE05C)
 val SpecialBlue = Color(0xFFB4D6FC)
 val SpecialBlue2 = Color(0xFF8ABEF8)
+val SpecialGray = Color(0xFFBBBBBB)
+val SpecialGray2 = Color(0xFF9B9B9B)
